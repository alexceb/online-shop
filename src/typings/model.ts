export interface ProductOption {
  color: string
  quantity: number
  power?: number[]
  storage?: string[]
}

export interface Product {
  id: number
  name: string
  brand: string
  price: string
  available: boolean
  weight: number
  options: ProductOption[]
}

export interface ProductConfiguration {
  color: string
  power?: number
  storage?: string
}

export interface CartItem extends ProductConfiguration{
  product: Product
  quantity: number
}
