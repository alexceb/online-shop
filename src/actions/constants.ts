export enum ActionNames {
  // Product list actions
  RECEIVED_DATA = 'RECEIVED_DATA',
  RECEIVED_DATA_ERROR = 'RECEIVED_DATA_ERROR',
  INCREASE_AMOUNT_IN_STOCK = 'INCREASE_AMOUNT_IN_STOCK',
  DECREASE_AMOUNT_IN_STOCK = 'DECREASE_AMOUNT_IN_STOCK',
  // Product details actions
  GET_PRODUCT_BY_ID = 'GET_PRODUCT_BY_ID',
  GET_PRODUCT_BY_ID_ERROR = 'GET_PRODUCT_BY_ID_ERROR',
  GET_PRODUCT_BY_ID_SUCCESS = 'GET_PRODUCT_BY_ID_SUCCESS',
  RESET_SELECTED_PRODUCT = 'RESET_SELECTED_PRODUCT',
  // Checkout actions
  REMOVE_ITEM_FROM_CART = 'REMOVE_ITEM_FROM_CART',
  INCREASE_AMOUNT = 'INCREASE_AMOUNT',
  DESCRESE_AMOUNT = 'DECREASE_AMOUNT',
  ADD_ITEM_TO_CART = 'ADD_ITEM_TO_CART',
  // Loader actions
  ENABLE_LOADER = 'ENABLE_LOADER',
  DISABLE_LOADER = 'DISABLE_LOADER',
}
