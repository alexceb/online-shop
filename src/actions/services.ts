import { api } from '../utils/api';

export function getInitialData() {
  return api.getProducts();
}