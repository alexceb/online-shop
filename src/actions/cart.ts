import { ActionNames } from './constants';
import { CartItem } from '../typings/model';

interface RemoveItemAction {
  type: typeof ActionNames.REMOVE_ITEM_FROM_CART
  payload: CartItem,
}

interface UpdateItemQuantity {
  type: typeof ActionNames.UPDATE_ITEM_QUANTITY
  payload: CartItem,
}

export type CartActions = 
  RemoveItemAction | 
  UpdateItemQuantity;

export const removeItemFromCart = (item: CartItem): RemoveItemAction => ({
  type: ActionNames.REMOVE_ITEM_FROM_CART,
  payload: item,
});

export const updateItemQuantity = (item: CartItem): UpdateItemQuantity => ({
  type: ActionNames.UPDATE_ITEM_QUANTITY,
  payload: item,
});
