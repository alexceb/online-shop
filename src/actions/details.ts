import { ActionNames } from './constants';
import { Product } from '../typings/model';
import { Dispatch } from 'redux';
import { api } from '../utils/api';

interface ProductResetAction {
  type: typeof ActionNames.RESET_SELECTED_PRODUCT
  payload: {
    isLoading: boolean,
  };
}

interface GetProductByIdFromApiAction {
  type: typeof ActionNames.GET_PRODUCT_BY_ID_SUCCESS
  payload: {
    products: Product[],
    product: Product,
    isLoading: boolean,
  }
}

interface GetProductByIdErrorAction {
  type: typeof ActionNames.GET_PRODUCT_BY_ID_ERROR
  payload: {
    error: Error,
    isLoading: boolean,
  }
}

export type DetailsActions = ProductResetAction |
                             GetProductByIdFromApiAction |
                             GetProductByIdErrorAction;

export const onGetProductByIdFromApi = (products: Product[], id: number) => {
  return (dispatch: Dispatch<DetailsActions>) => {
    return api.getProductById(id)
      .then(res => dispatch(onGetProductSuccess(products, res, id)))
      .catch(err => dispatch(onGetProductError(err)));
  }
}

export const onGetProductSuccess = (products: Product[], res: Product, id: number): GetProductByIdFromApiAction => ({
  type: ActionNames.GET_PRODUCT_BY_ID_SUCCESS,
  payload: {
    products: products.filter(product => product.id === id),
    product: res,
    isLoading: false,
  },
});

export const onGetProductError = (error: Error): GetProductByIdErrorAction => ({
  type: ActionNames.GET_PRODUCT_BY_ID_ERROR,
  payload: {
    error,
    isLoading: false,
  }
})

export const resetSelectedProduct = (): ProductResetAction  => ({
  type: ActionNames.RESET_SELECTED_PRODUCT,
  payload: { isLoading: true },
});
