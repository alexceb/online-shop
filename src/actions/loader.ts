import { ActionNames } from '../actions/constants';

interface EnableAction {
  type: typeof ActionNames.ENABLE_LOADER
  payload: boolean
}

interface DisableAction {
  type: typeof ActionNames.DISABLE_LOADER
  payload: boolean
}

export type LoaderActions = EnableAction | DisableAction;

export const enableLoader = (): EnableAction => ({
  type: ActionNames.ENABLE_LOADER,
  payload: true,
}); 

export const disableLoader = (): DisableAction => ({
  type: ActionNames.DISABLE_LOADER,
  payload: false,
});
