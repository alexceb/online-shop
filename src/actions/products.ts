import { Dispatch } from 'redux';

import { ActionNames } from './constants';
import { Product, CartItem } from '../typings/model';
import { api } from '../utils/api';

interface ReceivedDataAction {
  type: typeof ActionNames.RECEIVED_DATA
  payload: {
    data: Product[]
    isLoading: boolean
  };
}

interface ReceivedDataWithErrorAction {
  type: typeof ActionNames.RECEIVED_DATA_ERROR
  payload: {
    error: Error
    isLoading: boolean
  }
}

interface ChangeAmountAction {
  type: typeof ActionNames.INCREASE_AMOUNT_IN_STOCK | ActionNames.DECREASE_AMOUNT_IN_STOCK
  payload: {
    item: CartItem
    amount: number
  }
}

export type ProductActions = ReceivedDataAction | ChangeAmountAction | ReceivedDataWithErrorAction;

export const onGetData = () => {
  return (dispatch: Dispatch<ProductActions>) => {
    return api.getProducts()
      .then(res => dispatch(onGetDataSuccess(res)))
      .catch(error => dispatch(onGetDataError(error)));
  }
}

export const onGetDataSuccess = (data: Product[]): ReceivedDataAction => ({
  type: ActionNames.RECEIVED_DATA,
  payload: {
    data: data,
    isLoading: false,
  },
});

export const onGetDataError = (error: Error): ReceivedDataWithErrorAction => ({
  type: ActionNames.RECEIVED_DATA_ERROR,
  payload: {
    error,
    isLoading: false,
  },
});

export const increaseAmountInStock = (item: CartItem, amount: number): ChangeAmountAction => ({
  type: ActionNames.INCREASE_AMOUNT_IN_STOCK,
  payload: {
    item,
    amount,
  },
});

export const decreaseAmountInStock = (item: CartItem, amount: number): ChangeAmountAction => ({
  type: ActionNames.DECREASE_AMOUNT_IN_STOCK,
  payload: {
    item,
    amount,
  },
});
