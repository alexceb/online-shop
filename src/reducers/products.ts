import { ActionNames } from '../actions/constants';
import { ProductActions } from '../actions/products';
import { Product, CartItem, ProductOption } from '../typings/model';

export interface ProductsState {
  data: Product[]
  isLoading: boolean
}

const initialState: ProductsState = {
  data: [],
  isLoading: true,
};

const getConfiguredData = (
    data: Product[],
    payload: { item: CartItem, amount: number },
    increase: boolean
  ) => {

  let changedOptions;

  const changedQuantity = (value: number, increase: boolean, amount: number) => {
    return increase ? (value + amount) : (value - amount);
  };

  const getAvailability = (options: ProductOption[]): boolean => {
    return options.reduce((curr: number, next: ProductOption) => {
      if (next.quantity > 0) {
        return curr + next.quantity
      };
      return curr;
    }, 0) > 0;
  };

  return data
    .map(item => {
      if (item.id === payload.item.product.id) {
        changedOptions = item.options.map(option => {
          if (option.color === payload.item.color) {
            return {
              ...option,
              quantity: changedQuantity(option.quantity, increase, payload.amount),
            }
          }
          return option;
        });
        return {
          ...item,
          options: changedOptions,
          available: getAvailability(changedOptions),
        };
      }
      return item;
    })
}

export const productsReducer = (state = initialState, action: ProductActions): ProductsState => {
  let increase;
  switch (action.type) {
    case (ActionNames.RECEIVED_DATA):
      return {
        ...state,
        data: action.payload.data,
        isLoading: action.payload.isLoading,
      }
    case (ActionNames.RECEIVED_DATA_ERROR):
      return {
        ...state,
        isLoading: action.payload.isLoading,
      }
    case (ActionNames.INCREASE_AMOUNT_IN_STOCK):
      increase = true;
      return {
        ...state,
        data: getConfiguredData(state.data, action.payload, increase)
      }
    case (ActionNames.DECREASE_AMOUNT_IN_STOCK):
      increase = false;
      return {
        ...state,
        data: getConfiguredData(state.data, action.payload, increase)
      }
    default:
      return state;
  }
}
