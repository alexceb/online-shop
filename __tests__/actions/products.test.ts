import configureMockStore from 'redux-mock-store';
import thunk, { ThunkDispatch } from 'redux-thunk';

import { increaseAmountInStock, decreaseAmountInStock, ProductActions, onGetData } from '../../src/actions/products';
import { ActionNames } from '../../src/actions/constants';
import { CartItem, Product } from '../../src/typings/model';
import { products } from '../../src/utils/_DATA';

type DispatchExts = ThunkDispatch<ProductsState, void, ProductActions>;
type ProductsState = {
  data: Product[]
  isLoading: boolean
}

const middlewares= [thunk];
const mockStore = configureMockStore<ProductsState, DispatchExts>(middlewares);
const mockedData = [...products];
const store = mockStore({
  data: [],
  isLoading: true,
});


const mockedProduct: Product = {
  id: 8,
  name: "Samsung 40 UHD Smart TV",
  brand: "SAMSUNG",
  price: "3200",
  available: true,
  weight: 8.2,
  options: [
    {
      color: "black",
      quantity: 19,
    }
  ],
};
const mockedCartItem: CartItem = {
  product: {
    id: 1,
    name: "Philips hue bulb",
    brand: "Philips",
    price: "500",
    available: true,
    weight: 0.2,
    options: [
      {
        color: "white",
        power: [6.5, 9.5],
        quantity: 3,
      },
      {
        color: "red",
        power: [6.5, 9.5],
        quantity: 7,
      },
    ]
  },
  quantity: 3,
  color: 'white',
  subOption: {
    optionType: 'power',
    optionValue: '7',
  }
};

describe('Products page action creators', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('It dispatches RECEIVED_DATA action when getting initial data with products', () => {
    const expectedActions = [
      { type: ActionNames.RECEIVED_DATA, payload: { data: mockedData, isLoading: false, } },
    ];

    return store.dispatch(onGetData()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    })
  })

  it('It dispatches INCREASE_AMOUNT_IN_STOCK when decreasing cart item from checkout', () => {
    const input = mockedCartItem;
    store.dispatch(increaseAmountInStock(input, 2));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('It dispatches DECREASE_AMOUNT_IN_STOCK when increasing cart item from checkout', () => {
    const input = mockedCartItem;
    store.dispatch(decreaseAmountInStock(input, 1));
    expect(store.getActions()).toMatchSnapshot();
  });

});