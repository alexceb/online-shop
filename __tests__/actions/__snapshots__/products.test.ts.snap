// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products page action creators It dispatches DECREASE_AMOUNT_IN_STOCK when increasing cart item from checkout 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": 1,
      "item": Object {
        "color": "white",
        "product": Object {
          "available": true,
          "brand": "Philips",
          "id": 1,
          "name": "Philips hue bulb",
          "options": Array [
            Object {
              "color": "white",
              "power": Array [
                6.5,
                9.5,
              ],
              "quantity": 3,
            },
            Object {
              "color": "red",
              "power": Array [
                6.5,
                9.5,
              ],
              "quantity": 7,
            },
          ],
          "price": "500",
          "weight": 0.2,
        },
        "quantity": 3,
        "subOption": Object {
          "optionType": "power",
          "optionValue": "7",
        },
      },
    },
    "type": "DECREASE_AMOUNT_IN_STOCK",
  },
]
`;

exports[`Products page action creators It dispatches INCREASE_AMOUNT_IN_STOCK when decreasing cart item from checkout 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": 2,
      "item": Object {
        "color": "white",
        "product": Object {
          "available": true,
          "brand": "Philips",
          "id": 1,
          "name": "Philips hue bulb",
          "options": Array [
            Object {
              "color": "white",
              "power": Array [
                6.5,
                9.5,
              ],
              "quantity": 3,
            },
            Object {
              "color": "red",
              "power": Array [
                6.5,
                9.5,
              ],
              "quantity": 7,
            },
          ],
          "price": "500",
          "weight": 0.2,
        },
        "quantity": 3,
        "subOption": Object {
          "optionType": "power",
          "optionValue": "7",
        },
      },
    },
    "type": "INCREASE_AMOUNT_IN_STOCK",
  },
]
`;
